File: LeetCode/912/main.cpp
class Solution {
public:
  void merge(vector<int> &arr, int left, int mid, int right) {
    int i = left;
    int j = mid + 1;
    int k = 0;

    vector<int> temp(right - left + 1);

    while (i <= mid && j <= right) {
      if (arr[i] <= arr[j]) {
        temp[k++] = arr[i++];
      } else {
        temp[k++] = arr[j++];
      }
    }

    while (i <= mid) {
      temp[k++] = arr[i++];
    }

    while (j <= right) {
      temp[k++] = arr[j++];
    }

    // overwrite
    k = 0;

    for (int i = left; i <= right; i++) {
      arr[i] = temp[k++];
    }
  }

  void mergeSort(vector<int> &arr, int left, int right) {
    if (left >= right) {
      return;
    }

    int mid = left + (right - left) / 2;

    mergeSort(arr, left, mid);
    mergeSort(arr, mid + 1, right);
    merge(arr, left, mid, right);

    return;
  }

  vector<int> sortArray(vector<int> &arr) {

    mergeSort(arr, 0, arr.size() - 1);

    return arr;
  }
};

File: LeetCode/912/main.py
class Solution:
    def sortArray(self, arr: List[int]) -> List[int]:

        def merge(arr, left, mid, right):
            i, j = left, mid + 1
            ans = []

            while (i <= mid and j <= right):
                if (arr[i] < arr[j]):
                    ans.append(arr[i])
                    i += 1
                else:
                    ans.append(arr[j])
                    j += 1

            while (i <= mid):
                ans.append(arr[i])
                i += 1

            while (j <= right):
                ans.append(arr[j])
                j += 1

            # overwrite

            for i in range(len(ans)):
                arr[i + left] = ans[i]

            return arr

        def mergeSort(arr, left, right):
            if (left >= right):
                return

            mid = left + (right - left) // 2
            mergeSort(arr, left, mid)
            mergeSort(arr, mid + 1, right)
            ans = merge(arr, left, mid, right)

            return ans

        ans = mergeSort(arr, 0, len(arr) - 1)
        return ans
